### DISCUSSION TITLE
Serializing std::function

### DISCUSSION THREAD

**Discussion by tomek-szczesny** [nlohmann/json#2431] [Q&A] [answered]
Hi!

The title probably says it all: I've got a MIDI message filtering class with a std::function<bool(MidiMsg)> member, that can be either left blank (nullptr) or defined by a user.

A lengthy google and github research left me with no hope for doing this as c++ std::function serialization seems to be something out of reach.

If that is indeed the case, how could I possibly store this information?

Thanks in advance for any tips! :)

**Comment by nlohmann** [comment, marked as answer]
See https://github.com/nlohmann/json#arbitrary-types-conversions, and scroll to the entry for `boost::optional<T>`. This should be also applicable to `std::function<T>`.

**Reply by tomek-szczesny** [reply]
Thanks, I've given up that idea and found a workaround for this problem, but I'll try your suggestion anyway, out of sheer curiosity.
Given that there's no implementation of std::hash<std::function> neither is operator==(), how could this possibly be?

**Reply by nlohmann** [reply]
What do you mean?

**Reply by nlohmann** [reply]
Do you need further assistance with this issue?

**Reply by tomek-szczesny** [reply]
No, thank you, I abandoned the idea of serializing std::function long ago. Nevertheless, I haven't found any evidence that this would actually work and my attempts have failed (even though I did succeed with all sorts of other types).

As I stated above, I believe that in order to serialize an object, the object has to be "hashable" and comparable. I believe std::function has neither of these properties.

Currently I'm overwhelmed with other tasks and don't see myself testing further ideas, but if you feel that this is indeed doable, I'd be more than happy to see a working prototype or example code in README.md, or additional information what cannot be serialized, in contrast to the current claim that there are no limits.

**Reply by gregmarr** [reply]
It could be serialized as long as you have a way to convert a `std::function` to plain text and back again.  That might be something that requires reflection, or an external lookup table that stores a mapping between a particular `std::function` and a string.  That's not something that the library itself can provide, but it does provide you everything you need to do it yourself if you have that mapping.

>As I stated above, I believe that in order to serialize an object, the object has to be "hashable" and comparable. I believe std::function has neither of these properties.

I don't think that's true.  As far as I know, the only thing that needs to be hashable and comparable should be the key type for the map, which is locked to be a string.  There could be some area that I'm missing, but right now I can't think what that might be.

**Reply by tomek-szczesny** [reply]
So what you're saying is that I can map functions to something else and serialize that. This is not what I was after. I wanted a way to serialize directly an arbitrary std::function into a string, which on its own right a very complex problem that largely is not even attempted, if stackoverflow is to be believed.

**Reply by gregmarr** [reply]
This library doesn't have anything to help you with that to/from string part, as that's very application specific.  The general case of "convert this arbitrary function pointer into a string, write it to disk, and then read it back later and get back the same function in another run of the program" is impossible.  Limiting it to the same execution of the program makes it somewhat closer to possible, as long as you don't unload the dynamic module containing the function.  Having a registry when functions are registered at runtime with a unique persistent handle makes it much easier.  The OS can provide functions for getting a function pointer by name, but you still have the problem of how to get the name of the function.

Once you have that piece, the library can handle conversion of the `std::function<T>` to/from a JSON file when it's standalone or embedded in another handled type.  You need something like this, filling in the `...` section using the mapping method from above:

https://github.com/nlohmann/json#how-do-i-convert-third-party-types

    namespace nlohmann {
        template <typename T>
        struct adl_serializer<std::function<T>> {
            static void to_json(json& j, const std::function<T>& func) {
                if (func == nullptr) {
                    j = nullptr;
                } else {
                  j = ...;
                }
            }

            static void from_json(const json& j, std::function<T>& func) {
                if (j.is_null()) {
                    func = nullptr;
                } else {
                    opt = ...;
                }
            }
        };
    }

**Reply by nlohmann** [reply]
I agree with @gregmarr. If it would be possible to somehow translate a `std::function` from one machine into some sequence of bytes that could be translated on another computer back into the same `std::function`, then the piece of code above can be filled out with these translation functions and you have a `std::function` <-> JSON <-> `std::function` chain. 

The replacement of the `...` is left as an exercise to the reader ;-)

**Reply by tomek-szczesny** [reply]
I'm really glad we all agree.
